src/HealthChecks.Rabbitmq/RabbitMQHealthCheck.cs

```diff
@@ -26,49 +26,88 @@ public RabbitMQHealthCheck(RabbitMQHealthCheckOptions options)
     {
-        // TODO: cancellationToken unused, see https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/714
         try
         {
-            using var model = EnsureConnection().CreateModel();
-            return HealthCheckResultTask.Healthy;
+            var connection = await EnsureConnectionAsync(cancellationToken).ConfigureAwait(false);
+            await using var model = await connection.CreateChannelAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return HealthCheckResult.Healthy();
         }
         catch (Exception ex)
         {
-            return Task.FromResult(new HealthCheckResult(context.Registration.FailureStatus, exception: ex));
+            return new HealthCheckResult(context.Registration.FailureStatus, exception: ex);
         }
     }
 
-    private IConnection EnsureConnection()
-    {
-        _connection ??= _connections.GetOrAdd(_options, _ =>
-            {
-                var factory = _options.ConnectionFactory;
+    private async Task<IConnection> EnsureConnectionAsync(CancellationToken cancellationToken) =>
+        _connection ??= await _connections.GetOrAddAsync(_options, async options =>
+        {
+            var factory = options.ConnectionFactory;
 
-                if (factory is null)
+            if (factory is null)
+            {
+                Guard.ThrowIfNull(options.ConnectionUri);
+                factory = new ConnectionFactory
                 {
-                    Guard.ThrowIfNull(_options.ConnectionUri);
-                    factory = new ConnectionFactory
-                    {
-                        Uri = _options.ConnectionUri,
-                        AutomaticRecoveryEnabled = true
-                    };
+                    Uri = options.ConnectionUri,
+                    AutomaticRecoveryEnabled = true
+                };
 
-                    if (_options.RequestedConnectionTimeout is not null)
-                    {
-                        ((ConnectionFactory)factory).RequestedConnectionTimeout = _options.RequestedConnectionTimeout.Value;
-                    }
+                if (options.RequestedConnectionTimeout is not null)
+                {
+                    ((ConnectionFactory)factory).RequestedConnectionTimeout = options.RequestedConnectionTimeout.Value;
+                }
 
-                    if (_options.Ssl is not null)
-                    {
-                        ((ConnectionFactory)factory).Ssl = _options.Ssl;
-                    }
+                if (options.Ssl is not null)
+                {
+                    ((ConnectionFactory)factory).Ssl = options.Ssl;
                 }
+            }
 
-                return factory.CreateConnection();
-            });
+            return await factory.CreateConnectionAsync(cancellationToken).ConfigureAwait(false);
+        }).ConfigureAwait(false);
+}
 
-        return _connection;
+internal static class ConcurrentDictionaryExtensions
```

Comment:
It's not related to this PR, but it shows that we really need to get rid of that static cache, as it brings so much complexity and increases a risk for various bugs (leaks etc).

Reply:
Yes - agreed. I will try to make that change next.